server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -0,0 +1,249 @@
+    private final List<TaskResourceUsageTracker> taskResourceUsageTrackers;
+
+    private final TokenBucket taskCancellationRateLimiter;
+    private final TokenBucket taskCancellationRatioLimiter;
+
+    // Currently, only the state of SearchShardTask is being tracked.
+    // This can be generalized to Map<TaskType, SearchBackpressureState> once we start supporting cancellation of SearchTasks as well.
+    private final SearchBackpressureState state = new SearchBackpressureState();
+
+    public SearchBackpressureService(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool
+    ) {
+        this(
+            settings,
+            taskResourceTrackingService,
+            threadPool,
+            System::nanoTime,
+            List.of(
+                new NodeResourceUsageTracker(
+                    () -> ProcessProbe.getInstance().getProcessCpuPercent() / 100.0 >= settings.getNodeDuressCpuThreshold()
+                ),
+                new NodeResourceUsageTracker(
+                    () -> JvmStats.jvmStats().getMem().getHeapUsedPercent() / 100.0 >= settings.getNodeDuressHeapThreshold()
+                )
+            ),
+            Collections.emptyList()
+        );
+    }
+
+    public SearchBackpressureService(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool,
+        LongSupplier timeNanosSupplier,
+        List<NodeResourceUsageTracker> nodeResourceUsageTrackers,
+        List<TaskResourceUsageTracker> taskResourceUsageTrackers
+    ) {
+        this.settings = settings;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+        this.taskResourceTrackingService.addTaskCompletionListener(this);
+        this.timeNanosSupplier = timeNanosSupplier;
+        this.nodeResourceUsageTrackers = nodeResourceUsageTrackers;
+        this.taskResourceUsageTrackers = taskResourceUsageTrackers;
+
+        this.taskCancellationRateLimiter = new TokenBucket(
+            timeNanosSupplier,
+            getSettings().getCancellationRateNanos(),
+            getSettings().getCancellationBurst()
+        );
+
+        this.taskCancellationRatioLimiter = new TokenBucket(
+            state::getCompletionCount,
+            getSettings().getCancellationRatio(),
+            getSettings().getCancellationBurst()
+        );
+
+        threadPool.scheduleWithFixedDelay(this, getSettings().getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    public void run() {
+        try {
+            doRun();
+        } catch (Exception e) {
+            logger.debug("failure in search search backpressure", e);
+        }
+    }
+
+    public void doRun() {
+        if (getSettings().isEnabled() == false) {
+            return;
+        }
+
+        if (isNodeInDuress() == false) {
```

Comment:
Just thinking out loud - Do we track if the node comes out of duress after the task cancellations were done in the previous iteration?

Reply:
Do you think adding a `duress` boolean field in the stats API will help? Correlating when the node was in and out of duress, along with the number of cancellations that happened between that time frame might provide some insight. This would provide an instantaneous view of the cluster, so it might be more meaningful if there's an external monitoring process which constantly polls the stats API.

Alternatively, we can keep track of the number of cancellations since the node last went into duress.

Let's explore more on this in the follow up PRs for stats API.